### PR TITLE
Fix Lost update between last draft and sending a message

### LINKED ISSUE #8734: Lost update between last draft and sending a message
### Steps to reproduce

1. Open the app
2. Set up an account
3. Start composing a new message, pick the sender and add some body text
4. Wait until the draft is saved
5. Enter some more text into the body and click *Send* quickly

### Expected behavior

The last body modifications are saved and go into the message I send

### Actual behavior

My last modifications are lost

### Mail app version

3.4.0-beta.2

### Mailserver or service

_No response_

### Operating system

_No response_

### PHP engine version

None

### Web server

None

### Database

None

### Additional info

_No response_

### PR DESCRIPTION
**Pull request by hamza221**
fixes #8734

### CODE REVIEW

File: lib/Controller/DraftsController.php [hunk 1]
```diff
@@ -208,7 +208,7 @@ public function update(int $id,
 	 */
 	#[TrapError]
 	public function destroy(int $id): JsonResponse {
-		$message = $this->service->getMessage($id, $this->userId);
+		$message = $this->service->getMessage($id, $this->userId, LocalMessage::TYPE_DRAFT);
```
Comment (st3iny) [inline lib/Controller/DraftsController.php:211]: Nit: I'd prefer to use `\OCA\Mail\Service\OutboxService::getMessage()` to fetch outbox messages and `\OCA\Mail\Service\DraftsService::getMessage()` to fetch drafts.

This also applies to the other usages.

File: lib/Service/DraftsService.php [hunk 2]
```diff
@@ -204,11 +204,12 @@ public function sendMessage(LocalMessage $message, Account $account): void {
...
 	 * @param int $id
 	 * @param string $userId
+	 * @param int $type
 	 * @return LocalMessage
 	 * @throws DoesNotExistException
 	 */
-	public function getMessage(int $id, string $userId): LocalMessage {
-		return $this->mapper->findById($id, $userId, LocalMessage::TYPE_DRAFT);
+	public function getMessage(int $id, string $userId, int $type): LocalMessage {
+		return $this->mapper->findById($id, $userId, $type);
```
Comment (st3iny) [inline lib/Service/DraftsService.php:212]: Nit: Please revert (see above).

File: lib/Controller/OutboxController.php [hunk 3]
```diff
@@ -156,8 +155,8 @@ public function create(
 	 * @return JsonResponse
 	 */
 	#[TrapError]
-	public function createFromDraft(DraftsService $draftsService, int $id, int $sendAt = null): JsonResponse {
-		$draftMessage = $draftsService->getMessage($id, $this->userId);
+	public function createFromDraft(int $id, int $sendAt = null): JsonResponse {
+		$draftMessage = $this->service->getMessage($id, $this->userId);
```
Comment (hamza221) [inline lib/Controller/OutboxController.php:159]: on updateDraft() the message has .sentAt value so it gets a  TYPE_OUTGOING that's why we're using the outboxService here

Comment (hamza221) [reply lib/Controller/OutboxController.php:159]: So the new endpoint should be removed right? I can't see it used anywhere else

Comment (hamza221) [reply lib/Controller/OutboxController.php:159]: Or we to be more in line with https://github.com/nextcloud/mail/issues/8631 update draft shouldn't change message type even if it has a sentAt

Comment (ChristophWurst) [reply lib/Controller/OutboxController.php:159]: > on updateDraft() the message has .sentAt value so it gets a TYPE_OUTGOING

is what we should stop doing. The conversion has to happen explicitly, not on the basis of sendAt being set or not

File: lib/Controller/DraftsController.php [hunk 4]
```diff
@@ -175,9 +175,8 @@ public function update(int $id,
 		$message = $this->service->getMessage($id, $this->userId);
 		$account = $this->accountService->find($this->userId, $accountId);
 
-		($sendAt !== null)
-			? $message->setType(LocalMessage::TYPE_OUTGOING)
-			: $message->setType(LocalMessage::TYPE_DRAFT);
+	
+		$message->setType(LocalMessage::TYPE_DRAFT);
```
Comment (ChristophWurst) [inline lib/Controller/DraftsController.php:179]: `\OCA\Mail\Controller\DraftsController::create` still has the same logic and it is in use

1) Compose a new message
2) Send it before the first draft

There is a POST to create the *draft* but it results in implicit conversion to an outbox message.

Expectation is that the draft is created, then converted to an outbox message.

**Review by ChristophWurst — COMMENTED** [review]
:+1: lost update is no longer lost

but I found a new logic flaw :smiling_face_with_tear:

**Comment by hamza221** [comment]
The removed test doesn't make sense anymore in DraftServiceTest

**Comment by GretaD** [comment]
so, ready to get merged? :)

**Review by ChristophWurst — APPROVED** [review]
Messages that are sent right away and those that are a draft first are sent as expected :pray: :relaxed:

File: tests/Unit/Service/DraftsServiceTest.php [hunk 5]
```diff
@@ -373,54 +373,6 @@ public function testConvertToOutboxMessage(): void {
 		$this->draftsService->updateMessage($account, $message, $to, $cc, $bcc, $attachments);
 	}
 
-	public function testConvertToOutboxMessageNoRecipients(): void {
```
Comment (ChristophWurst) [inline tests/Unit/Service/DraftsServiceTest.php:376]: right, the test makes no sense anymore because we just dropped the implicit conversion :+1:

**Pull request merged**